pkg/netceptor/netceptor.go

```diff
@@ -197,38 +172,86 @@ func New(NodeID string, AllowedPeers []string) *Netceptor {
 		listenerRegistry:       make(map[string]*PacketConn),
 		sendRouteFloodChan:     nil,
 		updateRoutingTableChan: nil,
-		shutdownChans:          make([]chan bool, 6),
 		hashLock:               &sync.RWMutex{},
 		nameHashes:             make(map[uint64]string),
 		serviceAdsLock:         &sync.RWMutex{},
 		serviceAdsReceived:     make(map[string]map[string]*ServiceAdvertisement),
 		sendServiceAdsChan:     nil,
+		backendWaitGroup:       sync.WaitGroup{},
+		backendCount:           0,
 	}
 	s.reservedServices = map[string]func(*messageData) error{
 		"ping": s.handlePing,
 	}
+	if ctx == nil {
+		ctx = context.Background()
+	}
 	s.addNameHash(NodeID)
-	// shutdownChans[0] is for the main protocol loop
-	s.updateRoutingTableChan = tickrunner.Run(s.updateRoutingTable, time.Hour*24, time.Second*1, s.shutdownChans[1])
-	s.sendRouteFloodChan = tickrunner.Run(s.sendRoutingUpdate, RouteUpdateTime, time.Millisecond*100, s.shutdownChans[2])
-	s.sendServiceAdsChan = tickrunner.Run(s.sendServiceAds, ServiceAdTime, time.Second*5, s.shutdownChans[3])
-	go s.monitorConnectionAging(s.shutdownChans[4])
-	go s.expireSeenUpdates(s.shutdownChans[5])
+	s.context, s.cancelFunc = context.WithCancel(ctx)
+	s.updateRoutingTableChan = tickrunner.Run(s.context, s.updateRoutingTable, time.Hour*24, time.Second*1)
+	s.sendRouteFloodChan = tickrunner.Run(s.context, s.sendRoutingUpdate, RouteUpdateTime, time.Millisecond*100)
+	s.sendServiceAdsChan = tickrunner.Run(s.context, s.sendServiceAds, ServiceAdTime, time.Second*5)
+	go s.monitorConnectionAging()
+	go s.expireSeenUpdates()
 	return &s
 }
 
 // Shutdown shuts down a Receptor network protocol instance
 func (s *Netceptor) Shutdown() {
-	for i := range s.shutdownChans {
-		s.shutdownChans[i] <- true
-	}
+	s.cancelFunc()
 }
 
 // NodeID returns the local Node ID of this Netceptor instance
 func (s *Netceptor) NodeID() string {
 	return s.nodeID
 }
 
+// AddBackend adds a backend to the Netceptor system
+func (s *Netceptor) AddBackend(backend Backend, connectionCost float64) error {
+	sessChan, err := backend.Start(s.context)
+	if err != nil {
+		return err
+	}
+	s.backendWaitGroup.Add(1)
+	s.backendCount++
+	go func() {
+		defer s.backendWaitGroup.Done()
+		for {
+			select {
+			case sess, ok := <-sessChan:
+				if ok {
+					go func() {
+						err := s.runProtocol(sess, connectionCost)
+						if err != nil {
+							logger.Error("Backend error: %s\n", err)
+						}
+					}()
+				} else {
+					return
+				}
+			case <-s.context.Done():
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// DoneBackend signals to the wait group that the backend is done
+func (s *Netceptor) DoneBackend() {
```

Comment:
Is this really something we want to export? Im not sure I see the value of that, it seems scary

Reply:
Yep, this is a holdover from the old code and shouldn't exist any more